### ISSUE TITLE
webpack v4 migration guide and feature additions

### ISSUE THREAD

**Issue by TheLarkInn** [webpack/webpack.js.org#1706]
Hey docs team and all of our beloved contributors!!!! 

We are getting close to merge webpack/webpack#next (v4) into master soon and we have drafted a WIP change log and this means we can start laying the ground work and documentation for our migration guide and other changes!

Listed is the link to the changelog:

https://github.com/webpack/webpack/wiki/Changelog-WIP

I'd give you ownership to break up the work however you see fit! Let me know if the core team can help in any way!!!

🔥🔥Lots of exciting changes!🔥🔥

**Comment by reergymerej** [comment]
I've got a [PR in core] that modifies a new feature (`sideEffects` in package.json) and I'd like to make the sister change in the docs.  Can anyone point me in the right direction?

[PR in core]: https://github.com/webpack/webpack/pull/6314

**Comment by skipjack** [comment]
Hmmm, maybe in the __API__ section somewhere? Also target your PR to the `next` branch.

**Comment by reergymerej** [comment]
Thanks, @skipjack.  I searched the `next` branch and didn't see anything.  I may be the first to add it.

**Comment by skipjack** [comment]
> From #1795...

Create an upgrade path section with clearly defined, individual pages:

- Upgrade from v1 to v2
- Upgrade from v2 to v3
- Upgrade from v3 to v4

Each page should include any script helpers, and examples particular to it's version of the upgrade.

**Comment by skipjack** [comment]
I like the note from from #1795 but I think we should create a "Migrating" `group` under __Guides__ and title the articles simply `v1 - v2`, `v2 - v3`, etc.

**Comment by TheLarkInn** [comment]
Here is the first issue with a set of migration diffs. I think this is something we can run with? 

https://github.com/webpack/webpack.js.org/issues/1706

**Comment by TheLarkInn** [comment]
@skipjack, I've been starting to make a bunch of videos that are going to hopefully be embeddable into our documentation. I'm not sure where I can start to organize them. Would it work best to dump into a standalone issue?

**Comment by TheLarkInn** [comment]
> I like the note from from #1795 though in terms of actual wording/implementation, I think we should create a "Migrating" group under Guides and title the articles simply v1 - v2, v2 - v3, etc. (or 1.x.x to 2.x.x).

Has this already landed in next branch? If not, I will create an issue so that first time contributor can tackle just this work (creating the group).

**Comment by skipjack** [comment]
@jeremenichelli @TheLarkInn here's the start of the list of todos as I understand it:

- [ ] Rewrite plugin API section according to the new `tapable` interface and hooks (#1754)
- [ ] Finish hook descriptions and see if anything else is missing re plugin interface
- [ ] `CommonsChunkPlugin` removal and new `splitChunks` / `minimize` configuration opts (#1812)

@EugeneHlushko can you share some of the things you tackled? Once we get those, I'll move this list to the top-level comment so it's more visible and we can continue updating it.

**Comment by EugeneHlushko** [comment]
1. Profiling plugin you've referenced #1830
2. info-verbosity #1817 isnt really v4 its just a feature that landed already
3. nonce #1833 also isnt v4 but wasnt documented and is not in a great state as having dynamic nonce requires re-compilation and complex ssr setup, but at least it documents what it currently does.
4. custom hash function documented #1761

**Comment by skipjack** [comment]
Perfect, thanks @EugeneHlushko. Let us / @jeremenichelli know if there's anything else you notice that is missing from this list (whether it's already landed in `next` or needs to).

@jeremenichelli so you're thinking about closing this ticket and starting a new issue or pr that lists these out at the top? I think that makes sense to keep things clean.

**Comment by jeremenichelli** [comment]
Not closing it @skipjack, leave this one for Migration, and creating one to track all v4 Docs. Just redirecting and gardening directions so they don't derail and it's easier to track progress on docs 😊

**Comment by skipjack** [comment]
Makes sense, then maybe we re-title and tag it so it's clear it's specific to writing the migration guide.

**Comment by jeremenichelli** [comment]
Yeap, sounds good.

**Comment by nikunjg** [comment]
any updates in providing webpack 4 documentation

**Comment by jeremenichelli** [comment]
@nikunjg we are working on it 😊 you can see the progress in the conversations in the Issues and Pull requests and in the `next` branch where we are merging all the changes.

Sadly, we don't have the dates, but we hope to have it as soon as possible. Let us know if you would like to collaborate.

**Comment by nikunjg** [comment]
@jeremenichelli, that means documentation provided in the link https://webpack.js.org/guides/asset-management/ is for webpack 3. if yes, Should I use webpack 3 then?

**Comment by jeremenichelli** [comment]
Particularly for that section, I don't any issues with webpack v4 besides possible issues in the loaders mention there. I suggest you to go to the loader repository and see if it has been updated to work with v4.

**Comment by the-spyke** [comment]
Is there documentation for `optimization` section?

**Comment by the-spyke** [comment]
I'm upgrading an app to the 4.5 version from 3.11, and after a day I can say that a big chunk of the documentation is outdated or missing. Looks like `#0CJS` turned out to be `#0DJS`.
`¯\_(ツ)_/¯`

**Comment by xgqfrms-gildata** [comment]
# webpack 4

https://medium.com/webpack/webpack-4-released-today-6cdb994702d4

https://github.com/webpack/webpack/releases/tag/v4.0.0

**Comment by stephan-v** [comment]
No wonder you have such a bad reputation if you keep pulling stunts like these.

**Comment by Swivelgames** [comment]
With documentation being so imperative, you would think this would get a "beta" release until its been completed. I don't see how something could be released, specifically something whose usage depends on the education of its consumers, without the documentation. They go hand-in-hand with something like this.

Webpack is the only thing that does what it does; and it does it well. But documentation is imperative, especially with something configuration-based. It has and will be for the foreseeable future, be the one major crux of Webpack. And with as many changes that have come to Webpack in version 4, it's needed more now than in previous versions.

This should not have been released without the documentation. I'm not here to "bash" or "nay-say" the hard work that has been done by the Webpack team to put together an amazing version 4, but it's frustrating to the consumers of Webpack to be held back so significantly by something like documentation. To me, and I'm sure others may feel the same, launching without documentation (especially a Migration guide) is almost unacceptable :(

As we step closer to Month 4, I hope the documentation continues to chug along and find a completion date soon!

Keep up the good work guys.
